Completed Concept Entry for Javascript: Optional Chaining #5805 (#5826)
Codecademy/docs: content/javascript/concepts/optional-chaining/optional-chaining.md

---
Title: 'Optional Chaining'
Description: 'The optional chanining operator simplifies the comparison of multiple data properties in a chain of connected objects' 
Subjects: 
  - 'Web Development'
  - 'Computer Science'
Tags: 
  - 'Data'
  - 'Operators'
CatalogContent: 
  - 'introduction-to-javascript'
  - 'paths/front-end-engineer-career-path'
---

> 
> 

Reviewer: This is unnecessary space.

```suggestion
```

> The **optional chanining** operator simplifies the comparison of multiple data properties in a chain of connected objects. This is especially valuable if any of the properties are null as the operatior will return as undefined instead of throwing an error.

Reviewer: There is a spelling error, also, this sentence can be simplified as follows -

```suggestion
The **optional chaining** operator simplifies comparing multiple data properties in a chain of connected objects. This is especially valuable if any of the properties are `null`, as the operator will return `undefined` instead of throwing an error.
```

Optional chaning can be used to validate nested object properties, call optional functions or methods on optional objects, accessing arrary elements, and more. 

> The primary benefits of using optional chaining instead of regular syntax include simplifying the code to achieve the same result so it's easier to read, protecting against runtime errors and enhances maintainability.

Reviewer: "Enhances" can be changed to "enhancing" to maintain a parallel structure, and we should add a comma for clarity.

```suggestion
The primary benefits of using optional chaining instead of regular syntax include simplifying the code to achieve the same result, making it easier to read, protecting against runtime errors, and enhancing maintainability.
```

## Syntax

> The basic syntax for `optional chaining` can be defined as: 

Reviewer: 'optional chaining' is not a keyword or coding term, so it should not be enclosed in backticks.

```suggestion
The basic syntax for using optional chaining is as follows: 
```

```js
// To access an object property
object?.property

//To access an element in an array
array?.[index]

//To invoke a function (if it exists)
object?.method?.()
```

## Example
This person object lists an individual's details and containing properties, nested objects, an array, and a method. Optional chaining for each conditional evaluation is listed below. 

```js
const person = {
  name: "Alice",
  gender: "Female",
  age: 12,
  address: {
    street: "1111 Palm Ave", 
    city: "Broken Arrow",
    state: "Oklahoma"
  },
  favoriteFoods: ["pizza", "ice cream", "cake"], 
  commonPhrase: function() {
    return `${this.name} always says "I love ${this.address.state}."`;
  }
};
```

## Accessing Object Properties

> To search for the `state` object of `person`:

Reviewer: "Object" is misleading here, as `state` is a property of the nested address object.

```suggestion
To search for the `state` property of `person.address`:
```

```js
//Regular syntax
const state = person && person.address && person.address.state;
console.log(`State is: ${state}`);

//This can be rewritten as:
const state = person?.address?.state;

> console.log(`State is: ${state}`);

Reviewer: Add complete code, including definition for `person` object, so that the learner can quickly copy paste this in a code editor and run this code like this -

```
const person = {
  name: "Alice",
  gender: "Female",
  age: 12,
  address: {
    street: "1111 Palm Ave", 
    city: "Broken Arrow",
    state1: "Oklahoma"
  },
  favoriteFoods: ["pizza", "ice cream", "cake"], 
  commonPhrase: function() {
    return `${this.name} always says "I love ${this.address.state1}."`;
  }
};

// Regular syntax for checking if the state1 property exists
const state1Regular = person && person.address && person.address.state1;
console.log(`State (regular syntax) is: ${state1Regular}`);

// This can be rewritten as:
const state1Chaining = person?.address?.state1;
console.log(`State (optional chaining) is: ${state1Chaining}`);
```

Output:

```
State (regular syntax) is: Oklahoma
State (optional chaining) is: Oklahoma

```

//The output for both is: 
State is: Oklahoma

> ```

Reviewer: Include output separately like this - 

```
The output of the above code will be as follows:

```shell
State is: Oklahoma
```
```

## Accessing Arrary Elements
To search for the `cake` in the `favoriteFoods` array of `person`:

```js
//Regular Syntax
const food = person && person.favoriteFoods && person.favoriteFoods.find(item => item === "cake");
console.log(`Favorite Food is: ${food}`);

//This can be rewritten as:
const food = person?.favoriteFoods.find(item => item === "cake");

> console.log(`Favorite Food is: ${food}`)

Reviewer: Add complete code, including the definition for the `person` object, so that the learner can quickly copy and paste this into a code editor and run this code like this -

```
const person = {
  name: "Alice",
  gender: "Female",
  age: 12,
  address: {
    street: "1111 Palm Ave", 
    city: "Broken Arrow",
    state1: "Oklahoma"
  },
  favoriteFoods: ["pizza", "ice cream", "cake"], 
  commonPhrase: function() {
    return `${this.name} always says "I love ${this.address.state1}."`;
  }
};

// Regular Syntax for searching favorite food
const foodRegular = person && person.favoriteFoods && person.favoriteFoods.find(item => item === "cake");
console.log(`Favorite Food (regular syntax) is: ${foodRegular}`);

// This can be rewritten as:
const foodChaining = person?.favoriteFoods.find(item => item === "cake");
console.log(`Favorite Food (optional chaining) is: ${foodChaining}`);

```

Output:

```
Favorite Food (regular syntax) is: cake
Favorite Food (optional chaining) is: cake
```

//The output for both is: 

> Favorite Food is: cake

Reviewer: Include output separately like this - 

```
The output of the above code will be as follows:

```shell
Favorite Food is: cake
```
```

```js
//Regular Syntax
if (person && typeof person.commonPhrase === "function") {
  const phrase = person.commonPhrase();
console.log(`${phrase}`);
}

//This can be rewritten as:
const phrase = person?.commonPhrase?.();
console.log(`${phrase}`);

//The output for both is: 
Alice always says "I love Oklahoma."
```

## Codebyte Example

Run the following example and compare the regular syntax of conditional statements and optional chaining on a more complex data set.

```codebyte/js
const people = [
  {
    name: "Blake",
    gender: "Male",
    age: null,
    address: {
      street: "456 Maple Dr",
      city: "Austin",
      state: "Texas"
    }
  },
  {
    name: "Eve",
    gender: "Female",
    age: 15,
    address: {
      street: "789 Birch Ln",
      city: "",
      state: "Colorado"
    }
  },
  null,
  {
    name: "David",
    gender: "Male",
    age: 72,
    address: {
      street: "123 Acorn St",
      city: "Tampa",
      state: "Florida"
    }
  },
  {
    name: "Jospeh",
    gender: "Male",
    age: 9,
    address: {
      street: "987 Pine Ave",
      city: "Taos",
      state: "New Mexico"
    }
  }
];

> people.forEach((person, index) => {
>   const city = person && person.address && person.address.city;
>   console.log(`Person ${index + 1}: City: ${city}`);
> });
> 
> people.forEach((person, index) => {
> const age = person?.age;
> console.log(`Person ${index + 1}: Age: ${age}`);
> });

Reviewer: It would be better to directly specify that this is the "regular syntax" for conditional checks, and then show how optional chaining simplifies this:

```suggestion
// Regular syntax for checking if a person's city exists
people.forEach((person, index) => {
  const city = person && person.address && person.address.city;
  console.log(`Person ${index + 1}: City: ${city}`);
});

// Optional chaining for simpler access
people.forEach((person, index) => {
const age = person?.age;
console.log(`Person ${index + 1}: Age: ${age}`);
});
```

Reviewer: The output of the above code is as follows - 

```
Person 1: City: Austin
Person 2: City: 
Person 3: City: null
Person 4: City: Tampa
Person 5: City: Taos
Person 1: Age: null
Person 2: Age: 15
Person 3: Age: undefined
Person 4: Age: 72
Person 5: Age: 9
```

Ideally, the output of both syntaxes should be the same so that the learner understands that they can be used alternatively. Update the code to reflect the same.
